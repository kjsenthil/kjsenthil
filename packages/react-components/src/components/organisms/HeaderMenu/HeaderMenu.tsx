import React, { useState } from 'react';
import Img from 'gatsby-image';
import {
  Box,
  Button,
  ClearIcon,
  Divider,
  Drawer,
  Grid,
  Icon,
  InputAdornment,
  Link,
  ListItem,
  ListItemSecondaryAction,
  MenuIcon,
  Switcher,
  useMediaQuery,
  useTheme,
  Typography,
} from '../../atoms';
import { DisabledComponent, NavLink, SubHeader } from '../../molecules';
import { ClickAwayListener } from '../../particles';
import CoachingPopover from '../CoachingPopover';
import {
  Availability,
  CallBack,
  CallContainer,
  CashAndInvestGrid,
  Circle,
  CoachIcon,
  CoachIconContainer,
  CoachTextContainer,
  DrawerContainer,
  DrawerFooterContainer,
  IconGridDivider,
  InvestButton,
  LogoImage,
  ModalContainer,
  MyAccountsContainer,
  ProfileMenuItem,
  RotatedIcon,
  StyledAppBar,
  StyledCallBackLink,
  StyledMenuNavGrid,
  StyledMobileMenuList,
  StyledIconButton,
  StyledInputBase,
  StyledProfileList,
  StyledSubHeader,
  StyledToolbar,
  SwitcherLabel,
} from './HeaderMenu.styles';
import { StyledChildNavLink } from '../../molecules/NavLinkWithDropDown/NavLinkWithDropDown.styles';
import Spacer from '../../atoms/Spacer/Spacer';
import NavLinkWithDropDown, { NavLinkWithDropDownProps } from '../../molecules/NavLinkWithDropDown';
import CalendarIcon from './CalendarIcon';

type LinkWithSwitch = {
  onClick: (isEnabled: boolean) => void;
  type: 'switch';
};

type LinkWithPath = {
  path: string;
  type?: string;
};

type LinkType = LinkWithSwitch | LinkWithPath;

export type LinkData = {
  name: string;
  shouldShowInMainMenu: boolean;
  shouldShowInDrawer: boolean;
  color?: 'primary' | 'error';
  icon?: React.ReactElement;
  disabled?: boolean;
  type?: string;
  childLinks?: {
    path: string;
    name: string;
    disabled?: boolean;
  }[];
} & LinkType;

type GatsbyImage = {
  childImageSharp: {
    fluid: {
      aspectRatio: number;
      src: string;
      srcSet: string;
      sizes: string;
    };
  };
};

type CoachImages = {
  coachPortrait: GatsbyImage;
  coachIcon: GatsbyImage;
};

export interface HeaderMenuProps {
  homePath: string;
  links: Array<LinkData>;
  switchHandler: (evt: React.ChangeEvent<HTMLInputElement>) => void;
  isExpFeatureFlagEnabled?: boolean;
  currentUrl: string;
  isNonProd?: boolean;
  myAccountsUrl?: string;
  coachImages: CoachImages;
  showCoachPopover: boolean;
  toggleCalendlyModal: () => void;
  toggleCoachPopover: (value?: boolean) => void;
  navigate: (path: string) => {};
}

const HeaderMenu = ({
  currentUrl,
  homePath,
  links,
  isNonProd = true,
  switchHandler,
  isExpFeatureFlagEnabled = true,
  myAccountsUrl,
  coachImages,
  toggleCalendlyModal,
  showCoachPopover,
  toggleCoachPopover,
  navigate,
}: HeaderMenuProps) => {
  const theme = useTheme();
  const isMobile = useMediaQuery(theme.breakpoints.down('sm'));
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const [profileMenuOpen, setProfileMenuOpen] = useState<boolean>(false);
  const [menuOpen, setMenuOpen] = useState<boolean>(false);

  const handleProfileMenuClick = () => {
    setProfileMenuOpen(!profileMenuOpen);
  };

  const handleMenuClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
    setMenuOpen(true);
  };

  const handleMenuClose = () => {
    setAnchorEl(null);
    setMenuOpen(false);
  };

  const handlePhonelineStatus = () => {
    const today = new Date(Date.now());
    const day = today.getDay();
    const time = today.getHours();
    if (day === 0 || day === 6 || time >= 18 || time < 9) {
      return (
        <>
          <Circle data-testid="circle-phoneline-icon" color={theme.palette.error.main} />
          <Availability>We are currently unavailable</Availability>
        </>
      );
    }
    return (
      <>
        <Circle data-testid="circle-phoneline-icon" color={theme.palette.success.main} />
        <Availability>We are open, call us: 020 7189 9999</Availability>
      </>
    );
  };

  const renderSwitcher = (withInnerLabel: boolean, onClick: (evt) => void) => (
    <Switcher
      size="small"
      withInnerLabel={withInnerLabel}
      checked={isExpFeatureFlagEnabled}
      onClick={onClick}
    />
  );

  const navigateHome = () => navigate(homePath);

  const renderProfileMenuDropdownLinks = (profileLink: LinkData | undefined) =>
    profileLink?.childLinks?.map((childLink) => (
      <ProfileMenuItem
        tabIndex={0}
        key={`${childLink.name}-Profile-MenuItem`}
        onClick={() => navigate((childLink as LinkWithPath).path as string)}
      >
        <StyledChildNavLink isMobile={isMobile}>{childLink.name}</StyledChildNavLink>
      </ProfileMenuItem>
    ));

  const getOnLinkClick = (link: LinkData) =>
    link.type === 'switch'
      ? (link as LinkWithSwitch).onClick
      : () => navigate((link as LinkWithPath).path as string);

  const styledDivider = (
    <IconGridDivider>
      <Divider orientation="vertical" y={4} />
    </IconGridDivider>
  );

  const renderMyAccountsLink = () => (
    <MyAccountsContainer>
      <RotatedIcon name="upload" />
      <Link color="white" href={myAccountsUrl}>
        Back to old Bestinvest
      </Link>
    </MyAccountsContainer>
  );

  const renderCoachSignpost = () => (
    <CoachIconContainer>
      <ModalContainer>
        {showCoachPopover && (
          <ClickAwayListener mouseEvent="onMouseDown" onClickAway={() => toggleCoachPopover(false)}>
            <CoachingPopover
              image={
                <Img
                  fluid={coachImages.coachPortrait.childImageSharp.fluid}
                  alt="Portrait image of coach"
                />
              }
              onButtonClick={toggleCalendlyModal}
            />
          </ClickAwayListener>
        )}
      </ModalContainer>

      <CoachTextContainer>
        <Typography variant="sh1" color="inherit">
          YOUR COACH
        </Typography>
        <Link
          color="white"
          special
          onClick={() => (isMobile ? toggleCalendlyModal() : toggleCoachPopover())}
        >
          Book an appointment
        </Link>
      </CoachTextContainer>

      <CoachIcon coachIconUrl={coachImages.coachIcon.childImageSharp.fluid.src}>
        <CalendarIcon />
      </CoachIcon>
    </CoachIconContainer>
  );

  const disableableNavLinks = (children: React.ReactNode, index: number, disabled?: boolean) => {
    if (!disabled) {
      return <div key={`enabled-${index}`}>{children}</div>;
    }
    return (
      <DisabledComponent title="Coming soon" key={`disabled-${index}`}>
        {children}
      </DisabledComponent>
    );
  };

  const navLinkWithDropDown = (link: LinkData, mobile: boolean) => (
    <NavLinkWithDropDown
      key={`${link.name}-${mobile ? 'mobile' : 'desktop'}-menu-dropdown`}
      name={link.name}
      type={link.type as string}
      path={(link as LinkWithPath).path}
      childLinks={link.childLinks as NavLinkWithDropDownProps['childLinks']}
      selected={(link as LinkWithPath).path === currentUrl}
      navigate={navigate}
    />
  );

  const menuLink = (link: LinkData) => (
    <NavLink
      onClick={getOnLinkClick(link)}
      selected={(link as LinkWithPath).path === currentUrl}
      isMobile={isMobile}
      key={`${link.name}-desktop-menu`}
      variant="link"
    >
      {link.name}
    </NavLink>
  );

  const drawerLink = (link: LinkData) => (
    <>
      <ListItem
        key={`${link.name}-mobile-menu`}
        onClick={getOnLinkClick(link) as any}
        button={link.type === 'switch' ? false : undefined}
      >
        {!!link.icon && (
          <>
            {link.icon}
            <Spacer x={1} />
          </>
        )}

        <NavLink
          colorShade={link.color === 'error' ? undefined : 'dark2'}
          color={link.color ?? 'primary'}
          selected={(link as LinkWithPath).path === currentUrl}
          isMobile={isMobile}
          key={`${link.name}-mobile-link`}
          variant="link"
        >
          {link.name}
        </NavLink>

        {link.type === 'switch' && (
          <ListItemSecondaryAction key={`${link.name}-mobile-switch`}>
            {renderSwitcher(false, (evt) => getOnLinkClick(link)(evt.target.checked))}
          </ListItemSecondaryAction>
        )}
      </ListItem>
    </>
  );

  const renderDrawerList = () =>
    links
      .filter((link) => link.shouldShowInDrawer)
      .map((link, index) =>
        link.childLinks
          ? disableableNavLinks(navLinkWithDropDown(link, true), index, link.disabled)
          : disableableNavLinks(drawerLink(link), index, link.disabled)
      );

  const renderMenuNavLinks = () =>
    links
      .filter((link) => link.shouldShowInMainMenu)
      .map((link, index) => (
        <Grid
          item
          key={`${(link as LinkWithPath).path || link.name}-menu`}
          {...(link.disabled ? { disabled: true } : {})}
        >
          {link.childLinks
            ? disableableNavLinks(navLinkWithDropDown(link, false), index, link.disabled)
            : disableableNavLinks(menuLink(link), index, link.disabled)}
        </Grid>
      ));

  const renderCashAndInvestActions = (forSubMenu: boolean = true) => {
    const ButtonWrapper = ({ children }: { children: React.ReactElement }) =>
      forSubMenu ? (
        <Grid item container justifyContent="space-between" spacing={2} wrap="nowrap">
          {children}
        </Grid>
      ) : (
        children
      );
    return (
      <>
        <ButtonWrapper>
          <>
            <Grid item>
              <DisabledComponent title="Coming soon">
                <Button
                  data-testid="add cash"
                  color="primary"
                  wrap="nowrap"
                  startIcon={isMobile ? null : <Icon name="plus" />}
                  variant="outlined"
                >
                  Add cash
                </Button>
              </DisabledComponent>
            </Grid>

            <CashAndInvestGrid isMobile={isMobile} {...(isMobile ? { item: true, xs: 9 } : {})}>
              <DisabledComponent title="Coming soon">
                <InvestButton
                  data-testid="invest"
                  color="gradient"
                  variant="contained"
                  {...(isMobile
                    ? {}
                    : {
                        startIcon: <Icon name="investStatistics" />,
                      })}
                  isMobile={isMobile}
                >
                  Invest
                </InvestButton>
              </DisabledComponent>
            </CashAndInvestGrid>
          </>
        </ButtonWrapper>
      </>
    );
  };

  return (
    <Box data-testid="header">
      <StyledAppBar
        data-testid="header-menu"
        color="inherit"
        elevation={0}
        isMobile={isMobile}
        position="relative"
        style={{ zIndex: 1201 }}
      >
        <StyledToolbar variant="dense" isMobile={isMobile}>
          <Grid container justifyContent="space-between" alignItems="center">
            <Grid container justifyContent="space-between" direction="row" alignItems="center">
              <Grid item xs={2}>
                <Link onClick={navigateHome}>
                  <LogoImage isMobile={isMobile} />
                </Link>
              </Grid>

              {isMobile ? (
                <Grid item xs={9}>
                  <Grid container justifyContent="flex-end">
                    <StyledIconButton
                      edge="start"
                      color="primary"
                      aria-label={!menuOpen ? 'menu' : 'close menu'}
                      onClick={!menuOpen ? handleMenuClick : handleMenuClose}
                    >
                      {!menuOpen ? <MenuIcon /> : <ClearIcon />}
                    </StyledIconButton>

                    <Drawer
                      anchor="right"
                      keepMounted
                      open={Boolean(anchorEl)}
                      onClose={handleMenuClose}
                      style={{ zIndex: 1200 }} // Must be inline to override z-index set by MUI modal JS
                    >
                      <DrawerContainer>
                        {isExpFeatureFlagEnabled ? (
                          <StyledSubHeader>{renderCoachSignpost()}</StyledSubHeader>
                        ) : null}

                        <DisabledComponent title="Coming soon">
                          <StyledInputBase
                            placeholder="Markets, products, stocks"
                            startAdornment={
                              <InputAdornment position="start">
                                <Icon name="search" color="disabled" />
                              </InputAdornment>
                            }
                          />
                        </DisabledComponent>

                        <StyledMobileMenuList isExpFeatureFlagEnabled={isExpFeatureFlagEnabled}>
                          {isNonProd && (
                            <SwitcherLabel
                              labelPlacement="start"
                              control={<Box ml={1}>{renderSwitcher(true, switchHandler)}</Box>}
                              label="Experimental Features"
                            />
                          )}

                          {renderDrawerList()}
                        </StyledMobileMenuList>

                        <DrawerFooterContainer>
                          {renderCashAndInvestActions()}
                          <CallContainer>
                            {handlePhonelineStatus()}
                            <CallBack>
                              <DisabledComponent title="Coming soon">
                                <StyledCallBackLink>Request a callback</StyledCallBackLink>
                              </DisabledComponent>
                            </CallBack>
                          </CallContainer>
                        </DrawerFooterContainer>
                      </DrawerContainer>
                    </Drawer>
                  </Grid>
                </Grid>
              ) : (
                <Grid
                  item
                  container
                  direction="row"
                  justifyContent="space-between"
                  xs={10}
                  wrap="nowrap"
                >
                  <StyledMenuNavGrid item container justifyContent="flex-start" wrap="nowrap">
                    {renderMenuNavLinks()}
                  </StyledMenuNavGrid>

                  <Grid
                    item
                    container
                    spacing={2}
                    justifyContent="flex-end"
                    alignItems="center"
                    wrap="nowrap"
                    component="div"
                  >
                    {renderCashAndInvestActions(false)}

                    {styledDivider}

                    <Grid item>
                      <DisabledComponent title="Coming soon">
                        <Button
                          aria-controls="simple-menu"
                          aria-haspopup="true"
                          variant={undefined}
                          color="primary"
                          aria-label="list"
                          // onClick={handleProfileMenuClick} // To be added
                        >
                          <Icon name="list" />
                        </Button>
                      </DisabledComponent>
                    </Grid>

                    {styledDivider}

                    <Grid item>
                      <DisabledComponent title="Coming soon">
                        <Button
                          aria-controls="simple-menu"
                          aria-haspopup="true"
                          variant={undefined}
                          color="primary"
                          aria-label="search"
                          // onClick={handleProfileMenuClick} // To be added
                        >
                          <Icon name="search" />
                        </Button>
                      </DisabledComponent>
                    </Grid>

                    {styledDivider}

                    <Grid item>
                      <DisabledComponent title="Coming soon">
                        <Button
                          aria-controls="simple-menu"
                          aria-haspopup="true"
                          variant={undefined}
                          color="primary"
                          aria-label="shopping-cart"
                          // onClick={handleProfileMenuClick}  // To be added
                        >
                          <Icon name="cart" />
                        </Button>
                      </DisabledComponent>
                    </Grid>

                    {styledDivider}

                    <Grid item>
                      <Button
                        data-testid="profile"
                        aria-controls="simple-menu"
                        aria-haspopup="true"
                        variant={undefined}
                        color="primary"
                        aria-label="profile"
                        onClick={handleProfileMenuClick}
                      >
                        <Icon name="account" />
                      </Button>
                    </Grid>

                    {profileMenuOpen && (
                      <StyledProfileList profileMenuOpen={profileMenuOpen}>
                        {renderProfileMenuDropdownLinks(
                          links?.find((linkItem) => linkItem.name === 'Profile')
                        )}
                      </StyledProfileList>
                    )}
                  </Grid>
                </Grid>
              )}
            </Grid>
          </Grid>
        </StyledToolbar>

        {isExpFeatureFlagEnabled ? (
          <>
            {isMobile ? (
              !menuOpen && (
                <StyledSubHeader isMobile={isMobile}>
                  <Grid item container justifyContent="center">
                    {renderMyAccountsLink()}
                  </Grid>
                </StyledSubHeader>
              )
            ) : (
              <StyledSubHeader isMobile={isMobile}>
                <Grid container justifyContent="space-between" alignItems="center">
                  <Grid item>{renderMyAccountsLink()}</Grid>

                  <Grid item>
                    {isNonProd && (
                      <SwitcherLabel
                        labelPlacement="start"
                        control={<Box ml={1}>{renderSwitcher(true, switchHandler)}</Box>}
                        label="Experimental Features"
                      />
                    )}
                  </Grid>

                  <Grid item>{renderCoachSignpost()}</Grid>
                </Grid>
              </StyledSubHeader>
            )}
          </>
        ) : (
          !isMobile && (
            <SubHeader style={{ minHeight: '80px' }}>
              <Grid container justifyContent="space-between" alignItems="center">
                <Grid item>
                  {isNonProd && (
                    <SwitcherLabel
                      labelPlacement="start"
                      control={<Box ml={1}>{renderSwitcher(true, switchHandler)}</Box>}
                      label="Experimental Features"
                    />
                  )}
                </Grid>

                <Grid item>
                  <Button color="primary" variant="outlined" href={myAccountsUrl}>
                    My Accounts Login
                  </Button>
                </Grid>
              </Grid>
            </SubHeader>
          )
        )}
      </StyledAppBar>
    </Box>
  );
};

export default HeaderMenu;
