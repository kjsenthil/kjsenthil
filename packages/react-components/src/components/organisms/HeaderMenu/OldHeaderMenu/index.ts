import OldHeaderMenu, { OldHeaderMenuProps } from './OldHeaderMenu';

export type { OldHeaderMenuProps };
export default OldHeaderMenu;
