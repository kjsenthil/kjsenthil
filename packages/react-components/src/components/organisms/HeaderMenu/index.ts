import HeaderMenu, { HeaderMenuProps } from './HeaderMenu';

export type { HeaderMenuProps };
export default HeaderMenu;
