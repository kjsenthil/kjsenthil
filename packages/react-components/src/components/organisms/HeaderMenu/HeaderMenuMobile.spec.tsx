import React from 'react';
import { renderWithTheme, fireEvent, screen } from '@tsw/test-util';
import { navigate } from 'gatsby';
import { useMediaQuery } from '../../atoms';
import HeaderMenu from './HeaderMenu';

jest.mock('../../atoms', () => {
  const originalModule = jest.requireActual('../../atoms');

  return {
    ...originalModule,
    useMediaQuery: jest.fn(),
  };
});

jest.mock('gatsby', () => ({
  ...jest.requireActual('gatsby'),
  navigate: jest.fn(),
}));

const coachImages = {
  coachPortrait: {
    childImageSharp: {
      fluid: {
        aspectRatio: 0.65,
        src: 'coachPortrait.png',
        srcSet:
          '/coachPortrait.png 25w,/coachPortrait.png 50w,/coachPortrait.png 100w,coachPortrait.png 150w,/coachPortrait.png 200w,/coachPortrait.png 300w',
        sizes: '(max-width: 100px) 100vw, 100px',
      },
    },
  },
  coachIcon: {
    childImageSharp: {
      fluid: {
        aspectRatio: 0.65,
        src: 'coachIcon.png',
        srcSet:
          '/coachIcon.png 25w,/coachIcon.png 50w,/coachIcon.png 100w,coachIcon.png 150w,/coachIcon.png 200w,/coachIcon.png 300w',
        sizes: '(max-width: 100px) 100vw, 100px',
      },
    },
  },
};

const props = {
  homePath: '/',
  currentUrl: '/investments',
  switchHandler: () => {},
  coachImages,
  navigate,
  links: [
    {
      name: 'Investments',
      path: '/investments',
      shouldShowInDrawer: true,
      shouldShowInMainMenu: true,
      type: 'link',
      coachImages,
      childLinks: [
        {
          name: 'Stocks & Shares ISA',
          path: '/stocks-shares-isa',
        },
        {
          name: 'Investment accounts',
          path: '/investment-accounts',
        },
      ],
    },
    { name: 'Life plan', shouldShowInMainMenu: true, shouldShowInDrawer: true, path: '/life-plan' },
    { name: 'Documents', shouldShowInMainMenu: true, shouldShowInDrawer: true, path: '/documents' },
    {
      name: 'Help & Support',
      shouldShowInMainMenu: true,
      shouldShowInDrawer: true,
      path: '/help-and-support',
    },
    {
      name: 'Profile',
      shouldShowInMainMenu: false,
      shouldShowInDrawer: true,
      path: '/profile',
      type: 'link',
      childLinks: [
        {
          name: 'Logout',
          path: '/logout',
        },
      ],
    },
  ],
};

describe('Header Menu in Mobile', () => {
  beforeEach(() => {
    (useMediaQuery as jest.Mock).mockReturnValue('sm');
  });

  it('renders the header menu (with experimental features)', () => {
    const { result } = renderWithTheme(<HeaderMenu {...props} isExpFeatureFlagEnabled />);

    expect(result.container).toMatchSnapshot();
  });

  it('The link back to MyAccounts should be visible in the subheader', () => {
    const { result } = renderWithTheme(
      <HeaderMenu myAccountsUrl="https://google.com" isExpFeatureFlagEnabled {...props} />
    );
    const myAccountsLink = result.getByRole('link', { name: /Back to old Bestinvest/i });

    expect(myAccountsLink).toBeVisible();
    expect(myAccountsLink).toHaveAttribute('href', 'https://google.com');
  });

  it('The coach signpost should only be visible once the menu has been expanded', () => {
    const { result } = renderWithTheme(<HeaderMenu isExpFeatureFlagEnabled {...props} />);

    expect(result.getByText('YOUR COACH')).not.toBeVisible();
    expect(result.getByText('Book an appointment')).not.toBeVisible();

    fireEvent.click(result.getByRole('button', { name: 'menu' }));

    expect(result.getByText('YOUR COACH')).toBeVisible();
    expect(result.getByText('Book an appointment')).toBeVisible();
  });

  it('When the menu is expanded, clicking "Book an appointment" should not display the coaching modal, but should take the user to the appointments page', () => {
    const { result } = renderWithTheme(<HeaderMenu isExpFeatureFlagEnabled {...props} />);

    fireEvent.click(result.getByRole('button', { name: 'menu' }));

    expect(result.queryByText('Speak to a coach')).not.toBeInTheDocument();

    fireEvent.click(result.getByText('Book an appointment'));

    expect(result.queryByText('Speak to a coach')).not.toBeInTheDocument();
    expect(result.queryByRole('img')).not.toBeInTheDocument();

    expect(navigate).toHaveBeenCalledTimes(1);
    expect(navigate).toHaveBeenCalledWith('https://online.bestinvest.co.uk/bestinvest-plus#/');
  });

  it('should display a menu icon when menu is closed', () => {
    const { result } = renderWithTheme(
      <HeaderMenu myAccountsUrl="https://google.com" isExpFeatureFlagEnabled {...props} />
    );
    expect(result.getByLabelText('menu')).toBeVisible();
  });

  it('should display a close icon when menu is open', () => {
    const { result } = renderWithTheme(
      <HeaderMenu myAccountsUrl="https://google.com" isExpFeatureFlagEnabled {...props} />
    );
    fireEvent.click(
      result.getByRole('button', {
        name: /menu/i,
      })
    );

    expect(result.getByLabelText('close menu')).toBeVisible();
  });

  it('should not display child links of Nav Links', () => {
    const { result } = renderWithTheme(
      <HeaderMenu myAccountsUrl="https://google.com" isExpFeatureFlagEnabled {...props} />
    );

    expect(result.queryByText('Stocks & Shares ISA')).not.toBeInTheDocument();
    expect(result.queryByText('Investment accounts')).not.toBeInTheDocument();
  });

  it('should display child links of Nav Links when the arrow menu button is clicked', () => {
    const { result } = renderWithTheme(
      <HeaderMenu myAccountsUrl="https://google.com" isExpFeatureFlagEnabled {...props} />
    );

    fireEvent.click(
      result.getByRole('button', {
        name: /menu/i,
      })
    );

    fireEvent.click(screen.getByLabelText('open Investments menu'));

    expect(result.queryByText('Stocks & Shares ISA')).toBeInTheDocument();
    expect(result.queryByText('Investment accounts')).toBeInTheDocument();
  });

  it('should display a green icon when phone lines are open', () => {
    jest
      .spyOn(global.Date, 'now')
      .mockImplementation(() => new Date(`2021-08-24T13:01:58.135Z`).valueOf()); // Tuesday

    const { result } = renderWithTheme(
      <HeaderMenu myAccountsUrl="https://google.com" isExpFeatureFlagEnabled {...props} />
    );

    fireEvent.click(
      result.getByRole('button', {
        name: /menu/i,
      })
    );

    expect(result.queryByTestId('circle-phoneline-icon')).toHaveStyle('background-color: #3dd598');
  });

  it('should display the phone number when the phone line open', () => {
    jest
      .spyOn(global.Date, 'now')
      .mockImplementation(() => new Date(`2021-08-24T13:01:58.135Z`).valueOf()); // Tuesday

    const { result } = renderWithTheme(
      <HeaderMenu myAccountsUrl="https://google.com" isExpFeatureFlagEnabled {...props} />
    );

    fireEvent.click(
      result.getByRole('button', {
        name: /menu/i,
      })
    );

    expect(result.queryByText('We are open, call us: 020 7189 9999')).toBeInTheDocument();
    expect(result.queryByText('We are currently unavailable')).not.toBeInTheDocument();
  });

  it('should display unavailability text when the phone line closed', () => {
    Date.now = jest.fn().mockImplementation(() => new Date('2021-08-21T11:01:58.135Z').valueOf()); // Saturday

    const { result } = renderWithTheme(
      <HeaderMenu myAccountsUrl="https://google.com" isExpFeatureFlagEnabled {...props} />
    );

    fireEvent.click(
      result.getByRole('button', {
        name: /menu/i,
      })
    );

    expect(result.queryByText('We are currently unavailable')).toBeInTheDocument();
    expect(result.queryByText('We are open, call us: 020 7189 9999')).not.toBeInTheDocument();
  });

  it('should display a red icon when phone lines are closed - Weekend', () => {
    Date.now = jest.fn().mockImplementation(() => new Date('2021-08-21T11:01:58.135Z').valueOf()); // Saturday

    const { result } = renderWithTheme(
      <HeaderMenu myAccountsUrl="https://google.com" isExpFeatureFlagEnabled {...props} />
    );

    fireEvent.click(
      result.getByRole('button', {
        name: /menu/i,
      })
    );

    expect(result.queryByTestId('circle-phoneline-icon')).toHaveStyle('background-color: #eb3a17');
  });

  it('should display a red icon when phone lines are closed - Weekday, after 6pm', () => {
    Date.now = jest.fn().mockImplementation(() => new Date('2021-08-24T18:30:58.135Z').valueOf()); // Saturday

    const { result } = renderWithTheme(
      <HeaderMenu myAccountsUrl="https://google.com" isExpFeatureFlagEnabled {...props} />
    );

    fireEvent.click(
      result.getByRole('button', {
        name: /menu/i,
      })
    );

    expect(result.queryByTestId('circle-phoneline-icon')).toHaveStyle('background-color: #eb3a17');
  });
});
