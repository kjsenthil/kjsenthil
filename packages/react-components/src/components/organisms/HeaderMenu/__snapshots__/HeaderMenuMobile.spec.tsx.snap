// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Menu in Mobile renders the header menu (with experimental features) 1`] = `
.c2 {
  font-size: 0.8125rem;
  line-height: 1rem;
  -webkit-letter-spacing: 0.0125rem;
  -moz-letter-spacing: 0.0125rem;
  -ms-letter-spacing: 0.0125rem;
  letter-spacing: 0.0125rem;
  font-weight: 600;
  color: #6c00bf;
  white-space: break-spaces;
}

.c9 {
  font-size: 0.8125rem;
  line-height: 1rem;
  -webkit-letter-spacing: 0.0125rem;
  -moz-letter-spacing: 0.0125rem;
  -ms-letter-spacing: 0.0125rem;
  letter-spacing: 0.0125rem;
  font-weight: 600;
  color: #fff;
  white-space: break-spaces;
}

.c5 .MuiToolbar-root {
  border-top: 2px solid [object Object];
}

.c3 {
  background-image: url('/PurpleBestInvest.svg');
  background-repeat: no-repeat;
  background-size: contain;
  height: 37px;
  width: 124px;
  margin-left: 4px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-column-gap: 0.5rem;
  column-gap: 0.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c8 {
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
}

.c0 {
  min-height: 90px;
  margin: auto;
  max-width: 1440px;
  left: 0;
  right: 0;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
}

.c4 {
  padding-right: 4px;
}

.c6 {
  background: linear-gradient(267deg,#af61fe 20%,#7024fc 75%);
  color: #fff;
  padding-left: 2.25rem;
  padding-right: 2.5625rem;
  min-height: 34px;
}

.c1 {
  min-height: 90px;
}

@media (max-width:830px) {

}

@media (max-width:830px) {
  .c6 {
    padding: 0;
  }
}

@media (max-width:830px) {

}

<div>
  <div
    class="MuiBox-root MuiBox-root-1"
    data-testid="header"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionRelative MuiAppBar-colorInherit c0 MuiPaper-elevation0"
      data-testid="header-menu"
      style="z-index: 1201;"
    >
      <div
        class="MuiToolbar-root MuiToolbar-dense c1 MuiToolbar-gutters"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-space-between"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-content-xs-space-between"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
            >
              <button
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover c2 MuiLink-button MuiTypography-colorPrimary"
              >
                <div
                  class="c3"
                />
              </button>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-justify-content-xs-flex-end"
              >
                <button
                  aria-label="menu"
                  class="MuiButtonBase-root MuiIconButton-root c4 MuiIconButton-colorPrimary MuiIconButton-edgeStart"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiToolbar-root MuiToolbar-regular c5 c6 MuiToolbar-gutters"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-justify-content-xs-center"
        >
          <div
            class="c7"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root c8"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <title>
                upload
              </title>
              <path
                d="M17 9c1.389.027 2.213.15 2.826.62.208.16.395.346.554.554.59.77.619 2.323.62 4.496V15c0 2.346 0 4.018-.62 4.826-.16.208-.346.394-.554.554-.808.62-1.98.62-4.326.62h-7c-2.346 0-3.518 0-4.326-.62-.208-.16-.394-.346-.554-.554-.561-.73-.614-2.17-.62-4.174v-1.304c.006-2.004.059-3.443.62-4.174.16-.208.346-.394.554-.554C4.787 9.15 5.61 9.027 7 9c1.657 0 2 .448 2 1s-.343 1-2 1c-.412.008-.739.034-1.016.065-.255.029-.403.064-.488.093-.073.024-.097.043-.105.049-.07.053-.131.115-.184.184-.007.008-.024.032-.05.105-.028.085-.063.233-.092.488-.053.471-.063 1.455-.065 2.494v1.044c.002 1.04.012 2.023.065 2.494.029.255.064.403.093.488.025.073.043.097.049.105.053.07.115.131.184.184.008.006.032.024.105.049.085.029.233.064.488.093.55.062 1.297.065 2.516.065h7c1.22 0 1.966-.003 2.516-.065.255-.029.403-.064.488-.093.073-.024.097-.043.105-.049.07-.053.131-.115.184-.184.006-.008.024-.032.05-.105.028-.085.063-.233.092-.488.053-.471.063-1.455.065-2.494v-1.044c-.002-1.04-.012-2.023-.065-2.494-.029-.255-.064-.403-.093-.488-.024-.073-.043-.097-.049-.105-.053-.07-.115-.131-.184-.184-.008-.007-.032-.025-.105-.05-.085-.028-.233-.063-.488-.092-.277-.031-.604-.057-1.016-.065-1.657 0-2-.448-2-1s.343-1 2-1zm-5.603-5.798c.357-.27.85-.27 1.206 0l3.48 2.632c.485.366.534 1.076.105 1.505l-.089.089c-.346.346-.89.39-1.288.107L13 6.24v8.786c0 1.657-.448 2-1 2s-1-.343-1-2V6.24L9.19 7.535c-.398.284-.943.239-1.289-.107l-.089-.089c-.43-.43-.38-1.139.104-1.505z"
                fill-rule="evenodd"
              />
            </svg>
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineAlways c9 MuiTypography-colorPrimary"
            >
              Back to old Bestinvest
            </a>
          </div>
        </div>
      </div>
    </header>
  </div>
</div>
`;
